Modules/Scripted/WebServer/WebServer.py

```diff
@@ -0,0 +1,1507 @@
+        qt.QTimer.singleShot(0, self.onWriteableComplete)
+        self.connectionSocket.close()
+        self.logMessage('closed fileno %d' % (fileno))
+
+  def onServerSocketNotify(self,fileno):
+      self.logMessage('got request on %d' % fileno)
+      try:
+        (connectionSocket, clientAddress) = self.socket.accept()
+        fileno = connectionSocket.fileno()
+        self.requestCommunicators[fileno] = self.SlicerRequestCommunicator(connectionSocket, self.docroot, self.logMessage)
+        self.logMessage('Connected on %s fileno %d' % (connectionSocket, connectionSocket.fileno()))
+      except socket.error as e:
+        self.logMessage('Socket Error', socket.error, e)
+
+  def start(self):
+    """start the server
+    Uses one thread since we are event driven
+    """
+    try:
+      self.logMessage('started httpserver...')
+      self.notifier = qt.QSocketNotifier(self.socket.fileno(),qt.QSocketNotifier.Read)
+      self.logMessage('listening on %d...' % self.socket.fileno())
+      self.notifier.connect('activated(int)', self.onServerSocketNotify)
+
+    except KeyboardInterrupt:
+      self.logMessage('KeyboardInterrupt - stopping')
+      self.stop()
+
+  def stop(self):
+    self.socket.close()
+    if self.notifier:
+      self.notifier.disconnect('activated(int)', self.onServerSocketNotify)
+    self.notifier = None
+
+  def handle_error(self, request, client_address):
+    """Handle an error gracefully.  May be overridden.
+
+    The default is to print a traceback and continue.
+    """
+    print ('-'*40)
+    print ('Exception happened during processing of request', request)
+    print ('From', client_address)
+    import traceback
+    traceback.print_exc() # XXX But this goes to stderr!
+    print ('-'*40)
+
+  @classmethod
+  def findFreePort(self,port=2016):
+    """returns a port that is not apparently in use"""
+    portFree = False
+    while not portFree:
+      try:
+        s = socket.socket( socket.AF_INET, socket.SOCK_STREAM )
+        s.setsockopt( socket.SOL_SOCKET, socket.SO_REUSEADDR, 1 )
+        s.bind( ( "", port ) )
+      except socket.error as e:
+        portFree = False
+        port += 1
+      finally:
+        s.close()
+        portFree = True
+    return port
+
+
+
+#
+# WebServer logic
+#
+
+class WebServerLogic:
+  """Include a concrete subclass of SimpleHTTPServer
+  that speaks slicer.
+  """
+  def __init__(self, logMessage=None):
+    if logMessage:
+      self.logMessage = logMessage
+    self.port = 2016
+    self.server = None
+
+    moduleDirectory = os.path.dirname(slicer.modules.webserver.path.encode())
+    self.docroot = moduleDirectory + b"/Resources/docroot"
+
+  def start(self):
+    """Set up the server"""
+    self.stop()
+    self.port = SlicerHTTPServer.findFreePort(self.port)
+    self.logMessage("Starting server on port %d" % self.port)
+    self.logMessage('docroot: %s' % self.docroot)
+    # example: certfile = '/Users/pieper/slicer/latest/SlicerWeb/localhost.pem'
```

Comment:
Hi Steve. Here is some feedback:
The default pem file path and an "autostart" flag (telling if the WebServer should start automatically at application launch) could be read from application settings. Or autostart could be turned off by default and you could launch Slicer with specifying a small script that starts the WebServer with the parameters you like (--python-code "import WebServer; server=WebServer.... ; server.start(...)“.